Acinq

marcosrdz opened the pull request.

Overtorment commented on screen/lnd/browser.js:
```diff
@@ -229,190 +228,100 @@ export default class Browser extends Component {
   }
 
   renderWebView = () => {
-    if (Platform.OS === 'android') {
```
stop mixing different things in a single branch!
and whats with that webview that bothers you? are you familiar with "if it works - dont touch it"?
have you tested your webview change, both on ios and android? because I dont want to waste time testing stuff that wasnt broken.

Overtorment commented on class/acinqStrikeWallet.js:
```diff
@@ -0,0 +1,179 @@
+      throw new Error('API failure: ' + response.err + ' ' + JSON.stringify(response.originalResponse));
+    }
+
+    if (json && json.code && json.code === 401) {
```
wut

Overtorment commented on class/acinqStrikeWallet.js:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  /**
+   * Basically the same as this.getUserInvoices() but saves invoices list
```
incorrect jsdoc

Overtorment commented on class/acinqStrikeWallet.js:
```diff
@@ -0,0 +1,179 @@
+      body: { amount: Number(amount), description: String(description), currency: 'btc' },
+      auth: {
+        user: this.secret,
+        pass: '',
```
weird

marcosrdz replied:
Its how Frisbee handles auth for how Strike API asks for it.

Overtorment commented on screen/wallets/import.js:
```diff
@@ -74,6 +75,13 @@ export default class WalletsImport extends Component {
         await lnd.fetchTransactions();
         await lnd.fetchBalance();
         return this._saveWallet(lnd);
+      } else if (text.startsWith('sk_')) {
```
this is hacky, and might change without notice.
I suggest we will look for `strike://xxxxxxxxxxxxxxxxxxxx`

marcosrdz replied:
How would the user know to put in that schema? Strike doesn't show it that way afaik

Overtorment commented on screen/wallets/import.js:
```diff
@@ -74,6 +75,13 @@ export default class WalletsImport extends Component {
+        // ACINQ Strike
+        const wallet = new ACINQStrikeLightningWallet();
+        wallet.setSecret(text);
+        await wallet.authenticate();
```
not sure, whats gona happen if key is incorrect..? will it be handled gracefully?

marcosrdz replied:
Yup. they get the standard invalid info alert

Overtorment commented on class/acinqStrikeWallet.js:
```diff
@@ -0,0 +1,179 @@
+   *
+   * @param URI
+   */
+  setBaseURI(URI) {
```
we can get rid of it to simplify things. 
we are not going to support baseURI change by user. this is what this method was for in lndhub wallet

marcosrdz replied:
Fixed.

Overtorment commented:
Testing on emu.
When I open the strike wallet and pull to refresh, it throws error:

```
02-08 14:37:17.891  6025  6059 W ReactNativeJS: { [Error: Could not fetch balance from API: Error {"error":"Wallet false not found"}]
02-08 14:37:17.891  6025  6059 W ReactNativeJS:   line: 139301,
02-08 14:37:17.891  6025  6059 W ReactNativeJS:   column: 123,
02-08 14:37:17.891  6025  6059 W ReactNativeJS:   sourceURL: 'http://10.0.2.2:8081/index.delta?platform=android&dev=true&minify=false' }
02-08 14:37:17.913  6025  6059 I ReactNativeJS: 'Imported Strike', 'fetch balance took', 0.128, 'sec'
02-08 14:37:17.955  6025  6059 I ReactNativeJS: 'Imported Strike', 'fetch tx took', 0.042, 'sec'
02-08 14:37:17.969  6025  6059 I ReactNativeJS: wallets/transactions refreshFunction()
```

Overtorment commented:
`console.warn` throws obstructs emulator screen.
But everything works in general.

One thing: Acceptance test is missing.

Overtorment commented:
Another bug: expired invoices stay in the list as pending, and you can actually open them and see a qr code. Paying it will fail, obviously.

marcosrdz commented on class/acinqStrikeWallet.js:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    if (json && json.error) {
+      throw new Error('API error: ' + json.message + ' (code ' + json.code + ')');
```
It does. https://strike.acinq.co/documentation/api-reference

marcosrdz commented:
> Another bug: expired invoices stay in the list as pending, and you can actually open them and see a qr code. Paying it will fail, obviously.

Fixed.

marcosrdz commented:
> Testing on emu.
> When I open the strike wallet and pull to refresh, it throws error:
> 
> ```
> 02-08 14:37:17.891  6025  6059 W ReactNativeJS: { [Error: Could not fetch balance from API: Error {"error":"Wallet false not found"}]
> 02-08 14:37:17.891  6025  6059 W ReactNativeJS:   line: 139301,
> 02-08 14:37:17.891  6025  6059 W ReactNativeJS:   column: 123,
> 02-08 14:37:17.891  6025  6059 W ReactNativeJS:   sourceURL: 'http://10.0.2.2:8081/index.delta?platform=android&dev=true&minify=false' }
> 02-08 14:37:17.913  6025  6059 I ReactNativeJS: 'Imported Strike', 'fetch balance took', 0.128, 'sec'
> 02-08 14:37:17.955  6025  6059 I ReactNativeJS: 'Imported Strike', 'fetch tx took', 0.042, 'sec'
> 02-08 14:37:17.969  6025  6059 I ReactNativeJS: wallets/transactions refreshFunction()
> ```

Unable to replicate.

Overtorment commented on BlueComponents.js:
```diff
@@ -1037,6 +1038,10 @@ export class BlueTransactionListItem extends Component {
       if (isNaN(item.amount_satoshi)) {
         item.amount_satoshi = '0';
       }
+      const expectedExpiration = dayjs(this.props.item.updated + 86400000);
```
default expiration for Strike is 3600, I checked

Overtorment replied:
3600*1000

Overtorment commented on class/acinqStrikeWallet.js:
```diff
@@ -0,0 +1,160 @@
+    });
+
+    this.balance = 0;
+    this.user_charges_raw.forEach(charge => {
```
API returns default page size 30 elements. I imported my really old api key with a lot of charges, and it fetched only 30 latest. 
Im figuring thats  okay, lets not do complete iteration over all pages with a bunch of GETs, but once user starts using Strike in BW, he should not loose old transactions. I.e. if he starts using it a lot, regular `getUserCharges()` will always fetch only 30 latest, overwriting older ones.
Thus, somewhere here there should be a merge code: `this.user_charges_raw` should not be lost, newly fetched charges should be appended, any duplicates should be correctly handled (there should be no duplicates).

Overtorment commented:
BUGS:

* with that branch, "additional info" button disappears on regular LN invoices as well. its just not there
* Getting error `Error: Could not fetch balance from API: Error {"error":"Wallet false not found"}]` when pulling to refresh in a wallet. Maybe specific to my api key..?
* Is wallet card color the same as for regular LN wallet..?  You can do gradient from white to blue-ish on their website here https://strike.acinq.co

Overtorment commented:
Ok, the reason for Could not fetch balance is because you did not implement `fetchBalance`, and its being used from parent class when pulling to refresh.
You can make it a proxy to `fetchTransactions`

marcosrdz commented:
> Ok, the reason for Could not fetch balance is because you did not implement `fetchBalance`, and its being used from parent class when pulling to refresh.
> You can make it a proxy to `fetchTransactions`

Fixed.

marcosrdz commented:
> BUGS:
> 
> * with that branch, "additional info" button disappears on regular LN invoices as well. its just not there

FIXED
> * Getting error `Error: Could not fetch balance from API: Error {"error":"Wallet false not found"}]` when pulling to refresh in a wallet. Maybe specific to my api key..?

FIXED

> * Is wallet card color the same as for regular LN wallet..?  You can do gradient from white to blue-ish on their website here https://strike.acinq.co

https://github.com/BlueWallet/BlueWallet/issues/294 < @ncoelho

Overtorment commented:
Dont wait for #295, do your best with what you have.
Only tx merge is left

Overtorment commented on screen/wallets/list.js:
```diff
@@ -69,11 +63,11 @@ export default class WalletsList extends Component {
           try {
             let balanceStart = +new Date();
-            await BlueApp.fetchWalletBalances(that.lastSnappedTo || 0);
+            await BlueApp.fetchWalletBalances();
```
the meaning of this was to update only the wallet that is current. otherwise updatting all wallets and all their transactions might take looooong time

Overtorment replied:
so that was by design. not sure if its a good idea to change that.

Overtorment commented on class/lightning-custodian-wallet.js:
```diff
@@ -434,6 +434,7 @@ export class LightningCustodianWallet extends LegacyWallet {
 
     this._lastTxFetch = +new Date();
     this.transactions_raw = json;
+    return this.getTransactions();
```
it was supposed only to fetch.
outer code should use getter to get

Overtorment commented on class/acinqStrikeWallet.js:
```diff
@@ -86,8 +90,13 @@ export class ACINQStrikeLightningWallet extends LegacyWallet {
 
-  async getUserCharges() {
-    let response = await this._api.get('/charges');
+  async getUserCharges(page = 0) {
```
hows that supposed to work? when user reaches end of the list - it gona load new batch of transactions..? if yes then I dont see where `page` is passed.
also, with that approach balance of the wallet gona increase with scroll LOL

Overtorment commented on screen/wallets/transactions.js:
```diff
@@ -91,14 +105,17 @@ export default class WalletTransactions extends Component {
-      let txs = wallet.getTransactions();
-      for (let tx of txs) {
-        tx.sort_ts = +new Date(tx.received);
+      let txs = await wallet.fetchTransactions();
```
`refreshFunction` is only for screen redraw, not data fetch, as it says in jsdoc

Overtorment commented on screen/wallets/transactions.js:
```diff
@@ -91,14 +105,17 @@ export default class WalletTransactions extends Component {
-      for (let tx of txs) {
-        tx.sort_ts = +new Date(tx.received);
+      let txs = await wallet.fetchTransactions();
+      try {
```
try catch should not be here, nothing should fail here.

Overtorment commented:
closing in favor of https://github.com/BlueWallet/BlueWallet/pull/327 if Im not mistaken
